### DISCUSSION TITLE
I have ReferenceError: Cannot access 'o' before initialization  at Object.Z  when i build my project beause i imstance custom class in server components in Next 15

### DISCUSSION THREAD

**Discussion by ki-ned** [vercel/next.js#72757] [Help] [answered]
### Summary

In my project I created a main class apiService with different methods.
I've created child classes that inherit from it, in development everything works fine but as soon as I do a build I get this error:
ReferenceError: Cannot access ‘o’ before initialization
at Object.Z

I get this error wherever I create a new apiService () or a child in server component to use the methods defined in the class

page.tsx 
![image](https://github.com/user-attachments/assets/ea66601e-c725-4b06-8070-ab3ead0e2df1)

child CaisseService 
![image](https://github.com/user-attachments/assets/5aff7456-65a9-4f5a-aa85-d1906f70ed18)

apiService
![image](https://github.com/user-attachments/assets/bdbd876e-49fc-471f-86eb-f3a69e827233)

**Comment by icyJoseph** [comment]
Oh that doesn't look good...

Let's see, could you add this to your `next.config` file?

```js
  swcMinify: true,
  experimental: {
    serverMinification: false,
  },
```

What do you see if you do that?

**Reply by ki-ned** [reply]
this is what I get 

 Collecting page data  .Error: Failed to collect configuration for /operation
    at <unknown> (C:\Users\kined\node_modules\next\dist\build\utils.js:1131:23)
    at async Span.traceAsyncFn (C:\Users\kined\node_modules\next\dist\trace\trace.js:153:20) {    
  [cause]: ReferenceError: Cannot access 'ApiService' before initialization
      at Object.A (C:\Users\kined\.next\server\chunks\407.js:2930:48)
      at 85406 (C:\Users\kined\.next\server\chunks\407.js:2871:51)
      at __webpack_require__ (C:\Users\kined\.next\server\webpack-runtime.js:25:43)
      at 7994 (C:\Users\kined\.next\server\chunks\407.js:2933:82)
      at __webpack_require__ (C:\Users\kined\.next\server\webpack-runtime.js:25:43)
      at 87555 (C:\Users\kined\.next\server\app\(main)\operation\page.js:284:19)
      at Function.__webpack_require__ (C:\Users\kined\.next\server\webpack-runtime.js:25:43)
      at async getLayoutOrPageModule (C:\Users\kined\node_modules\next\dist\server\lib\app-dir-module.js:37:15)
      at async collectAppPageSegments (C:\Users\kined\node_modules\next\dist\build\segment-config\app\app-segments.js:50:45)

**Reply by icyJoseph** [reply]
Ok, so, it wasn't the code minify that was the issue.

I wonder, if during build the files are not loaded/executed in the order you expect them to, if at all?

Could you do some console.log debugging, to see in which order things are called at build time? It is possible for pages to be built by different workers, and that could, perhaps cause some unexpected code execution?

**Reply by icyJoseph** [reply]
Take a look here, https://github.com/vercel/next.js/discussions/71944#discussioncomment-11074298, for a possible way to make a single worker take care of things.

**Reply by ki-ned** [reply]
Ok, Thanhs a lot

**Reply by ki-ned** [reply]
I looked, set experimental.workerThreads to false and experimental.cpus to 1.
No change ApiService is called before its declaration in the compile files

**Reply by icyJoseph** [reply]
Alright interesting... 🤔 

I wonder then, if we have circular imports somehow? https://stackoverflow.com/questions/68315826/referenceerror-cannot-access-before-initialization-circular-dependency

Cannot access before initialization is mostly experienced in the Temporal Dead Zone, IIRC:

```js
{
  // Temporal dead zone 
  console.log(foo); // ReferenceError: Cannot access 'foo' before initialization
  let foo = 10;
}
```

Copy that into your console, including the `{` and `}`, and that'll yell at you...

So somehow, when building for production, that is happening.

Let's discard SWC, you can do:

```js
 swcMinify: false,
```

I think... I don't remember if that works in recent versions.

**Reply by ki-ned** [reply]
That's what happens in the compiled files, you'd have to change the order of execution and I don't think I have any circular imports that would need to be revirified.

**Reply by icyJoseph** [reply]
Did you try with Babel instead of SWC?

**Reply by ki-ned** [reply]
No, I haven't tried with Babel

**Reply by ki-ned** [reply, marked as answer]
I've got the solution, thanks for your help. All I had to do was fire a dynamic import of my class into my server component.

//import CaisseService from "@/src/services/app/caisse.service";
import RecuValide from "./RecuValide";
import { CaisseInterface } from "@/types/app_types";

export default async function page({
  params,
}: {
  params: Promise<{ uuid?: string }>;
}) {
  const uuid = (await params).uuid;

  if (!uuid) return <></>;

   const myClass = (await import("@/src/services/app/caisse.service")).default;
  const caisse = new myClass();
  const data: { data?: CaisseInterface } | null | undefined = await caisse.get_(
    { id: uuid }
  );

  if (!data) return <></>;

  return <RecuValide data={data?.data} />;
}

**Comment by ki-ned** [comment]
use dynamiic import in your sever component :
const myClass = (await import("@/src/services/app/caisse.service")).default;
const caisse = new myClass();
